### ISSUE TITLE
item lost and frozen during dragging

### ISSUE THREAD

**Issue by panzi** [yahoo/react-dnd-touch-backend#31]
When I start dragging an item the item freezes after dragged for a few pixels and instead the viewport scrolls with the item being frozen in the middle of the viewport. When I then drag the item again it works. Here is a video that illustrates the issue: https://www.youtube.com/watch?v=1SNyTLEQaE4

I of course use a custom drag layer and I render a simplified version of the dragged item there (no interaction code, only view). Only the number is the drag handle. Using the HTML5 backend on Desktop works exactly as expected.

Sadly I don't have a physical device to test this at the moment, only the android emulator. But it behaves the same in the Chrome device toolbar and the Firefox OS simulator. My host OS Fedora Linux 23.

**Comment by longlho** [comment]
do u have a simplified sample code to demon this?

**Comment by panzi** [comment]
Sorry, no. But I found out when I don't call `connectDragSource` (the number) and `connectDropTarget` (the whole item) on the dragged element when `isDragging == true` and also change the `<div>` inside the drag handle (the number element) to a `<span>` (and make it `display: block` again) everything works just fine. But it has to be both changes. One alone does not work. Absolutely strange.

**Comment by panzi** [comment]
I also tried to write my own minimal touch backend and it had the exact same problem! I still use my own touch backend because it also supports Windows Phone through MS Pointer Events. Will probably publish that backend next week.

**Comment by longlho** [comment]
Yeah I did some digging around w/ similar situation, since touch evt propagation is different and it interferes w/ scrolling also https://github.com/yahoo/react-dnd-touch-backend/issues/26#issuecomment-208098139 Doing this isn't as straightfoward as HTML5 DnD API

**Comment by knuterik91** [comment]
@longlho  Im having sort of the same issues as @panzi. 
I have no problem sorting elements inside one list, however when I try to drag an element from one list to another list then the item freezes with touch backend, no issues with html5 backend. 

I have created a repo with this issue [https://github.com/knuterik91/react-dnd-touch-html5-example](https://github.com/knuterik91/react-dnd-touch-html5-example)

**Comment by evanc** [comment]
I'm also seeing this, I think it's happening when the thing you're dragging unmounts (which works just fine in the HTML5 backend)

**Comment by longlho** [comment]
PR's welcomed 😃 (see #50)

**Comment by evanc** [comment]
I spent several hours debugging this today, if I find the issue I will certainly PR it.

**Comment by evanc** [comment]
I think it's actually because browsers stop sending `touchmove` events if the element the `touch` started on goes away- whether it's the thing that was actually listening to touch events or not. Weird.

**Comment by nvanselow** [comment]
In case this helps anyone else: I was able to solve this issue (dragging stops when a DOM element is removed/unmounts) by creating an "element cache" DOM node. This solution is still a work in progress. I will try to update this answer as it solidifies. I'm not sure if the touch backend should do this automatically or it should be handled by the developer when this specific case comes up.

In `beginDrag`, I call a function that copies the DOM element to a DOM node that will not be removed.

``` javascript
// -- Drag Source --
beginDrag(props, monitor, component) {
  props.cacheDragElement(component); // <- copies the DOM element, passed in from parent described below
  return {
    // the props you need
  };
},

endDrag(props, monitor) {
  // passed in function that clears out the elementCache div so it is empty for the next drag
  props.clearDragElementCache();
}
```
``` javascript
// -- Parent component: ---
constructor(props) {
   // other constructor setup
    
   this.elementCache = null; // stores the ref to the element cache
   this.cacheDragElement = this.cacheDragElement.bind(this);
   this.clearDragElementCache = this.clearDragElementCache.bind(this);
}

cacheDragElement(dragComponent) {
  if (this.elementCache && dragComponent) {
    // remember to `import {findDOMNode} from 'react-dom';` at the top of your file
    findDOMNode(this.elementCache).appendChild(findDOMNode(dragComponent));
  }
}

clearDragElementCache() {
  if (this.elementCache) {
    let elementCacheInDom = findDOMNode(this.elementCache);
      while (elementCacheInDom.hasChildNodes()) {
        elementCacheInDom.removeChild(elementCacheInDom.lastChild);
      }
  }
}

render () {
   <div className="dragContainer">
       <div></div> // The jsx that renders the components

       <div className="elementCache" ref={elementCache => this.elementCache = elementCache} />
   </div>  
}
```

In the css, I've set the `elementCache` to `display: none`.

This solution keeps a copy of the DOM with the touch event active so that dragging can continue even if the React code has unmounted the component.

**Comment by mszczepanczyk** [comment]
In my case I've managed to make it work by replacing conditional rendering with div hiding using styles.

So instead of

```javascript
render () {
  return connectDragSource(connectDropTarget((
    <div className='wrapper'>
      {props.isOver
        ? <div className='target' />
        : <div className='source' />
      }
    </div>
  )))
}
```

I have

```javascript
render () {
  return connectDragSource(connectDropTarget((
    <div className='wrapper'>
      <div className='target' style={{display: props.isOver ? 'block' : 'none'}} />
      <div className='source' style={{display: !props.isOver ? 'block' : 'none'}} />
    </div>
  )))
}
```

**Comment by NineThou** [comment]
In my case dragging item was frozen when the container component which was connected to the touch backend was rerendering

**Comment by eugeneford** [comment]
Also applied this fix in my project. Any considerations if this would be applied to backend itself?

**Comment by eugeneford** [comment]
I would try

**Comment by Finec** [comment]
In our case, when dragging too fast, the dragEnd event didn't fire (on Chrome 72), so the drag item freeze on the screen.

Here's a temporary fix for the wrapping container.

`handleDragStart = () => {
    window.addEventListener('mouseup', this.handleMouseUp);
  }

  handleMouseUp = () => {
    window.dispatchEvent(new Event('dragend'));    
    window.removeEventListener('mouseup', this.handleMouseUp);
  }

  componentWillUnmount() {
    window.removeEventListener('dragstart', this.handleDragStart);
  }

  componentDidMount() {
    window.addEventListener('dragstart', this.handleDragStart);
}`

**Comment by Finec** [comment]
In our case, when dragging too fast, the dragEnd event didn't fire (on Chrome 72), so the drag item freeze on the screen.

Here's a temporary fix for the wrapping container:

```javascript
handleDragStart = () => {
  window.addEventListener('mouseup', this.handleMouseUp);
}

handleMouseUp = () => {
  window.dispatchEvent(new Event('dragend'));    
  window.removeEventListener('mouseup', this.handleMouseUp);
}

componentWillUnmount() {
  window.removeEventListener('dragstart', this.handleDragStart);
}

componentDidMount() {
  window.addEventListener('dragstart', this.handleDragStart);
}
```
